runtime/drivers/clickhouse/olap.go

```diff
@@ -206,22 +208,124 @@ func (c *connection) AlterTableColumn(ctx context.Context, tableName, columnName
+		create.WriteString(" ORDER BY ")
+		create.WriteString(prop)
+	} else if engine == "MergeTree" {
+		// need ORDER BY for MergeTree
+		// it is optional for other engines
+		create.WriteString(" ORDER BY tuple() ")
+	}
+
+	// partition_by
+	if prop, ok := tableOpts["partition_by"].(string); ok {
+		create.WriteString(" PARTITION BY ")
+		create.WriteString(prop)
+	}
+
+	// primary_key
+	if prop, ok := tableOpts["primary_key"].(string); ok {
+		create.WriteString(" PRIMARY KEY ")
+		create.WriteString(prop)
+	}
+
+	// sample_by
+	if prop, ok := tableOpts["sample_by"].(string); ok {
+		create.WriteString(" SAMPLE BY ")
+		create.WriteString(prop)
+	}
+
+	// ttl
+	if prop, ok := tableOpts["ttl"].(string); ok {
+		create.WriteString(" TTL ")
+		create.WriteString(prop)
+	}
+
+	// settings
+	if prop, ok := tableOpts["settings"].(string); ok {
+		create.WriteString(" SETTINGS ")
+		create.WriteString(prop)
+	}
+
+	// write sql query
+	create.WriteString(" AS ")
+	create.WriteString(sql)
 	return c.Exec(ctx, &drivers.Statement{
-		Query:    fmt.Sprintf("CREATE OR REPLACE TABLE %s ENGINE = MergeTree ORDER BY tuple() AS %s", safeSQLName(name), sql),
+		Query:    create.String(),
 		Priority: 100,
 	})
 }
 
 // InsertTableAsSelect implements drivers.OLAPStore.
 func (c *connection) InsertTableAsSelect(ctx context.Context, name, sql string, byName, inPlace bool, strategy drivers.IncrementalStrategy, uniqueKey []string) error {
-	return fmt.Errorf("clickhouse: data transformation not yet supported")
+	if !inPlace {
+		return fmt.Errorf("clickhouse: inserts does not support inPlace=false")
+	}
+	if strategy == drivers.IncrementalStrategyAppend {
+		return c.Exec(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("INSERT INTO %s %s", safeSQLName(name), sql),
+			Priority:    1,
+			LongRunning: true,
+		})
+	}
+
+	if strategy == drivers.IncrementalStrategyMerge {
+		return c.WithConnection(ctx, 1, true, false, func(ctx context.Context, ensuredCtx context.Context, _ *dbsql.Conn) error {
+			// create temporary table
+			tmp := uuid.New().String()
+			err := c.Exec(ctx, &drivers.Statement{
+				Query: fmt.Sprintf("CREATE TABLE %s ENGINE=MergeTree ORDER BY tuple() AS %s", safeSQLName(tmp), sql),
+			})
+			if err != nil {
+				return err
+			}
+			defer func() {
+				_ = c.Exec(ctx, &drivers.Statement{
+					Query: fmt.Sprintf("DROP TABLE IF EXISTS %s", safeSQLName(tmp)),
+				})
+			}()
+
+			// Drop the rows from the target table where the unique key is present in the temporary table
+			cols := strings.Join(uniqueKey, ",")
+			err = c.Exec(ctx, &drivers.Statement{
+				Query: fmt.Sprintf("DELETE FROM %s WHERE (%s) IN (SELECT %s FROM %s)", safeSQLName(name), cols, cols, safeSQLName(tmp)),
+			})
+			if err != nil {
+				return err
+			}
+
+			// Insert the new data into the target table
+			return c.Exec(ctx, &drivers.Statement{
+				Query: fmt.Sprintf("INSERT INTO %s SELECT * FROM %s", safeSQLName(name), safeSQLName(tmp)),
+			})
```

Comment:
I'm fine with keeping this if you think it has use cases, but just FYI we don't need to support all strategies for all engines. I'm wondering if given ClickHouse has `ReplacingMergeTree` (even with its quirks of eventual consistency), we should just recommend people use that instead? Because ClickHouse doesn't have transactions, right? So there might be queries in between deleting the old data and inserting the new?

Reply:
Okay. I thought we want to keep consistency with respect to supported strategies. In the other case I think its better to remove this since lightweight deletes are also experimental in clickhouse.

> So there might be queries in between deleting the old data and inserting the new

From what I understand these lightweight deletes are stored separately and table is not changed immediately. For select queries the deleted rows are handled on the fly and eventually a background migration fixes the table atomically. But given these deletes happen in background there are lot of considerations to think of like these can also fail and clickhouse keeps complaining about a failed migration on subsequent queries. I think its best to remove them for now.

Reply:
Okay, let's remove it then. With both deletes and inserts having eventual consistency, there could be cases where a query is missing data.